org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ExtractInterfaceHandler.java

```diff
@@ -0,0 +1,132 @@
+import org.eclipse.lsp4j.CodeActionParams;
+import org.eclipse.ltk.core.refactoring.Refactoring;
+import org.eclipse.ltk.core.refactoring.participants.ProcessorBasedRefactoring;
+
+public class ExtractInterfaceHandler {
+
+	public static CheckExtractInterfaceResponse checkExtractInterfaceStatus(CodeActionParams params) {
+		ICompilationUnit cu = JDTUtils.resolveCompilationUnit(params.getTextDocument().getUri());
+		if (cu == null) {
+			return null;
+		}
+		IType type = cu.findPrimaryType();
+		if (type == null) {
+			return null;
+		}
+		CodeGenerationSettings settings = PreferenceManager.getCodeGenerationSettings(cu);
+		ExtractInterfaceProcessor processor = new ExtractInterfaceProcessor(type, settings);
+		IMember[] extractableMembers;
+		try {
+			extractableMembers = processor.getExtractableMembers();
+		} catch (JavaModelException e) {
+			return null;
+		}
+		Member[] members = Arrays.stream(extractableMembers).map(member -> {
+			try {
+				if (member instanceof SourceMethod) {
+					SourceMethod method = ((SourceMethod) member);
+					return new Member(method.getElementName(), Signature.getSignatureSimpleName(method.getReturnType()),
+							Arrays.stream(method.getParameterTypes()).map(parameterType -> Signature.getSignatureSimpleName(parameterType)).toArray(String[]::new), member.getHandleIdentifier());
+				} else if (member instanceof SourceField) {
+					SourceField field = ((SourceField) member);
+					return new Member(field.getElementName(), Signature.getSignatureSimpleName(field.getTypeSignature()), field.getHandleIdentifier());
+				} else {
+					return null;
+				}
+			} catch (JavaModelException e) {
+				return null;
+			}
+		}).filter(member -> member != null).toArray(Member[]::new);
+		return new CheckExtractInterfaceResponse(members, type.getElementName());
+	}
+
+	public static Refactoring getExtractInterfaceRefactoring(CodeActionParams params, List<String> handleIdentifiers, String interfaceName) {
+		ICompilationUnit cu = JDTUtils.resolveCompilationUnit(params.getTextDocument().getUri());
+		if (cu == null) {
+			return null;
+		}
+		IType type = cu.findPrimaryType();
+		if (type == null) {
+			return null;
+		}
+		CodeGenerationSettings settings = PreferenceManager.getCodeGenerationSettings(cu);
+		ExtractInterfaceProcessor processor = new ExtractInterfaceProcessor(type, settings);
+		try {
+			IMember[] selectedMembers = Arrays.stream(processor.getExtractableMembers()).filter((member) -> {
+				return handleIdentifiers.contains(member.getHandleIdentifier());
+			}).toArray(IMember[]::new);
+			processor.setExtractedMembers(selectedMembers);
+			processor.setTypeName(interfaceName);
+			processor.setReplace(JavaLanguageServerPlugin.getPreferencesManager().getPreferences().getExtractInterfaceReplaceEnabled());
+			processor.setAnnotations(true);
+			Refactoring refactoring = new ProcessorBasedRefactoring(processor);
+			if (refactoring.checkFinalConditions(new NullProgressMonitor()).isOK()) {
```

Comment:
We could add a log. For instance,
```
Refactoring refactoring = new ProcessorBasedRefactoring(processor);
RefactoringStatus status = refactoring.checkFinalConditions(new NullProgressMonitor());
if (status.isOK()) {
	return refactoring;
} else {
	JavaLanguageServerPlugin.logError(status.toString());
}
```
You can try to use `Pet` as an interface name. It will just fail.

Reply:
Yeah, we can do that. Besides, I would like to add a check request to avoid these known fail scenarios, for other unknown scenarios, Let's add a log here.